#!/usr/bin/env Rscript

main <- function() {
  ###################### LIBRARIES ###################### 
  ## install bioconductor installer, if needed
  if (!requireNamespace("BiocManager")) install.packages("BiocManager")
  ## install and load required R packages
  if(!require(Biostrings)) BiocManager::install("Biostrings", dependencies = TRUE)
  library(Biostrings)
  if(!require(ggplot2)) install.packages("ggplot2", dependencies = TRUE)
  library(ggplot2)
  if(!require(msa)) BiocManager::install("msa", dependencies = TRUE)
  library(msa)
  if(!require(ggmsa)) BiocManager::install("ggmsa", dependencies = TRUE)
  library(ggmsa)
  if(!require(seqinr)) install.packages("seqinr", dependencies = TRUE)
  library(seqinr)
  
  ###################### ARGPARSE ###################### 
  
  ## parse arguments
  args <- commandArgs(trailingOnly = TRUE)
  ## extended blast file generated by `2_blast_qc.sh`
  chr_categories <- args[1]  
  data_dir <- dirname(chr_categories)
  setwd(data_dir)
  # option for output filename 
  if (is.na(args[2])==FALSE){
    output_file <- args[2]
  } else {
    output_file <- paste0(data_dir, '/chr_compare.csv')
  }
  
  ###################### MAIN ###################### 

  ## import blast output file 
  ext_blast_df <- read.csv(chr_categories, header=T)
  ext_blast_df$gene = gsub(".+ - (\\w+) - .+", "\\1", 
                                ext_blast_df$match_description)
  ## write file containing only mismatched info, if mismatches exist
  ext_blast_df_mismatched = ext_blast_df[which(
    ext_blast_df$hit_v_orf=="incomplete_hit"), ]
  if (nrow(ext_blast_df_mismatched) > 0){
    write.csv(ext_blast_df_mismatched, file=paste0(data_dir,"/chr_mismatched.csv"), 
              row.names = FALSE)
  }
  
  # prepare data for plotting matches & mismatches
  ext_blast_df$num_match = 1
  ext_blast_df.collapsed = aggregate(
    num_match ~ gene+hit_v_orf, data=ext_blast_df, FUN=sum)

  ## plot hits 
  myPalette <- c("#984EA3", "#FFFB00")
  p = ggplot(data=ext_blast_df.collapsed, aes(x=gene, y=num_match, fill=hit_v_orf)) +
    geom_bar(stat="identity", position="stack", color="grey")  +
    theme_classic(base_size = 28) +
    labs(y="BLASTn Hits", x="") +
    theme(axis.text.x = element_text(angle=315, hjust=0.2, vjust=0.5),
          plot.title = element_text(hjust = 0.5),
          panel.grid.major.y = element_line(
            size = 0.5, linetype = 'solid', colour = "grey"),
          plot.margin = unit(c(0.4, 0.4, 0.4, 0.4), "cm"),
          legend.position = "right", legend.title=element_text(size=22),
          legend.text=element_text(size=22),
          legend.background = element_rect(
            fill="white", size=0.5, linetype="solid", colour ="black")) +
    scale_fill_manual("BLASTn \nPerformance", values = myPalette) 
  # save plot to SVG file
  img_file = paste0(data_dir, "/03_chr_performance.svg") 
  ggsave(
    img_file,
    plot = p,
    width = 5, height = 4)
  
  ## gather information for doing an optional MSA
  # collapse all identical proteins for MSA
  ext_blast_df$num_hits = 1
  ext_blast_df.collapsed = aggregate(
    num_hits ~ gene+mutation_type, data=ext_blast_df, FUN=sum)
  # return nt seq corresponding to best_aaseq
  best_ntseq_v = as.character()
  for (row in 1:nrow(ext_blast_df)) {
    if (ext_blast_df$best_aaseq[row]==ext_blast_df$og_aaseq[row]){ # if best prot is blast hit
      best_ntseq_v = append(best_ntseq_v, ext_blast_df$ntseq[row])
    } else {
      best_ntseq_v = append(best_ntseq_v, ext_blast_df$orfseq[row])
    }
  }
  ext_blast_df$best_ntseq = best_ntseq_v # append row of ntseq corresponding to best_ntseq
  ext_blast_df.msa = aggregate(
    num_hits ~ gene+mutation_type+best_ntseq+best_aaseq, data=ext_blast_df, FUN=sum)
  ext_blast_df.msa$mutation_type = as.character(ext_blast_df.msa$mutation_type)
  ext_blast_df.collapsed.khs.mut = ext_blast_df.msa[which(
    ext_blast_df.msa$mutation_type=="mutant" & ext_blast_df.msa$gene=="KHS"),]
  ext_blast_df.collapsed.khr.mut = ext_blast_df.msa[which(
    ext_blast_df.msa$mutation_type=="mutant" & ext_blast_df.msa$gene=="KHR"),]
  
  ## write files for MSA to determine KHR/KHS protein functionality
  ext_blast_df.msa.khs = ext_blast_df.msa[which(ext_blast_df.msa$gene=="KHS"),]
  write.csv(ext_blast_df.msa.khs, file=paste0(data_dir,"/KHS_msa_tbl.csv"), 
            row.names = FALSE)
  ext_blast_df.msa.khr = ext_blast_df.msa[which(ext_blast_df.msa$gene=="KHR"),]
  write.csv(ext_blast_df.msa.khr, file=paste0(data_dir,"/KHR_msa_tbl.csv"), 
            row.names = FALSE)
  
  ## cladogram of mutants (excludes "truncated" proteins)
  library("ape")
  library("Biostrings")
  library("ggplot2")
  library("ggtree")
  nwk <- system.file("extdata", "sample.nwk", package="ggtree")
  tree <- read.tree(nwk)
  tree
  # create neighbor joining tree
  #library('phangorn')
  #library('TreeTools')
  # convert KHR msa to fasta, then read in fasta
  tox_table=read.csv("/Users/angela/rowley/dsrna-survey/chr_toxins/output/Liti/KHR-msa.csv")
  tox_table=tox_table[which(tox_table$mutation_type!="truncated"),]
  tox_table$fasta_header=paste0(
    ">", tolower(treeseqdata$gene), "_mutant_", 1:nrow(treeseqdata), 
    " (", tox_table$num_hits, " strains)")
  Xfasta <- character(nrow(tox_table) * 2) # create object to store fasta data
  Xfasta[c(TRUE, FALSE)] <- tox_table$fasta_header
  Xfasta[c(FALSE, TRUE)] <- tox_table$best_aaseq
  writeLines(Xfasta, "toxin_msa_input.fasta")
  # start MSA by importing fasta file
  mySequences <- readAAStringSet("toxin_msa_input.fasta")
  toxinAln=msa(mySequences, type="protein", method="Muscle")
  writeLines(toxinAln, "toxin_msa_output.fasta")
  write.phyDat(x=toxinAln, format="phylip", file="toxin_msa_output.fasta")
  
  class(toxinAln) <- "AAMultipleAlignment"
  ggmsa(toxinAln, color = "Clustal", start=221, stop=280, 
        font = "DroidSansMono", char_width = 0.5, seq_name = TRUE )
  ggmsa(toxinAln, 221, 280, seq_name = TRUE, char_width = 0.5) + 
    geom_seqlogo(color = "Chemistry_AA") + geom_msaBar()
  
  
  toxinAln2 <- msaConvert(toxinAln, type="seqinr::alignment") # convert to seqinr alignment format
  d <- dist.alignment(toxinAln2, "identity")
  toxinTree <- nj(d) # construct tree using jeighbor joining algorithm
  msaPrettyPrint(toxinAln, output="pdf", file="tox_msa.pdf", showNames="left", 
                 shadingMode="similar", shadingColors="blues", 
                 showLogo="none", askForOverwrite=FALSE)
  plot(toxinTree, main=paste("Phylogenetic Tree of", treeseqdata$gene[1], "Toxin Sequences"))
  ## different tree method
  toxinAln_AAbin = msaConvert(toxinAln, type="ape::AAbin")
  toxinAln_phyDat = msaConvert(toxinAln, type="phangorn::phyDat")
  toxinAln_phyDat = as.phyDat(toxinAln_AAbin)
  
  read.phyDat(file, format = "phylip", type = "DNA", ...)
  NJTree(toxinAln_AAbin)
  
  treeseqdata=ext_blast_df.msa.khr[which(
    ext_blast_df.msa.khr$mutation_type!="truncated"),]
  treeseqdata$nstrains_label=paste(treeseqdata$num_hits,"strains")
  # order data so mutants are named in order of most common mutant first (ex:"khr_mutant_1")
  treeseqdata <- treeseqdata[order(treeseqdata$num_hits, decreasing = TRUE), ] 
  treeseqdata$mutant_id=paste0(tolower(treeseqdata$gene), "_mutant_", 1:nrow(treeseqdata))
  treeseqdata$aabin=as.AAbin(strsplit(treeseqdata$best_aaseq, ""))
  mydata = phyDat(treeseqdata$aabin, type="AA")
  
  NJTree()

  
  ## plot hits 
  myPalette <- c("#377EB8", "#4DAF4A", "#FA3232", "#984EA3")
  p = ggplot(data=ext_blast_df.collapsed, aes(x=gene, y=num_hits, fill=mutation_type)) +
    geom_bar(stat="identity", position="stack", color="grey")  +
    theme_classic(base_size = 28) +
    labs(y="BLASTn Hits", x="") +
    theme(axis.text.x = element_text(angle=315, hjust=0.2, vjust=0.5),
          plot.title = element_text(hjust = 0.5),
          panel.grid.major.y = element_line(
            size = 0.5, linetype = 'solid', colour = "grey"),
          plot.margin = unit(c(0.4, 0.4, 0.4, 0.4), "cm"),
          legend.position = "right", legend.title=element_text(size=22),
          legend.text=element_text(size=22),
          legend.background = element_rect(
            fill="white", size=0.5, linetype="solid", colour ="black")) +
    scale_fill_manual("Mutation Type", values = myPalette)
  # save plot to SVG file
  img_file = paste0(data_dir, "/03_chr_mutations.svg") 
  ggsave(
    img_file,
    plot = p,
    width = 5, height = 4)

}
main()